### PR TITLE
Support checkpoints

### PR DESCRIPTION
**Pull request by karthikprasad**
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs change / refactoring / dependency upgrade

## Motivation and Context / Related issue
#373 

## How Has This Been Tested (if it applies)
Unit tests.

## Checklist

- [x] The documentation is up-to-date with the changes I made.
- [x] I have read the **CONTRIBUTING** document and completed the CLA (see **CONTRIBUTING**).
- [x] All tests passed, and additional code has been covered with new tests.

### CODE REVIEW

File: opacus/grad_sample/grad_sample_module.py [hunk 1]
```diff
@@ -425,6 +425,18 @@ def rearrange_grad_samples(
 
         return activations, backprops
 
+    def state_dict(self, *args, **kwargs):
```
Comment (ffuuugor) [inline opacus/grad_sample/grad_sample_module.py:428]: Let's reopen this conversation: do we want to reroute *all* unknown (i.e. not explicitly declared in GradSampleModule) attributes to the underlying `_module`?
Right now we only do this for named parameters (see `__getattr__`) .

The reason I'm raising this here is because with new `state_dict` and `load_state_dict` we create a discrepancy between the output of `state_dict` and the output of `named_parameters()` method - which can produce unexpected results.
I think at least we should also override all `named_*` method here too, and at most - think about being more radical and just updating `__getattr__`.

Full context: I was planning to override `named_*` method for other reasons, just didn't get to publishing a PR yet

File: opacus/privacy_engine.py [hunk 2]
```diff
@@ -493,3 +494,50 @@ def get_epsilon(self, delta):
             Privacy budget (epsilon) expended so far.
         """
         return self.accountant.get_epsilon(delta)
+
+    def save_checkpoint(
+        self,
+        *,
+        path: Union[str, os.PathLike, BinaryIO, IO[bytes]],
+        module: GradSampleModule,
+        optimizer: DPOptimizer,
```
Comment (ffuuugor) [inline opacus/privacy_engine.py:503]: Should we maybe make optimizer optional here? 

1) You actually only need to save/load optimizer state if you're using parameter scheduling. Otherwise it should be enough to save model and the accountant.

2) Talking about schedulers, what's the recommended way to save/load if you're using noise scheduler? Do we want people to save it by themselves or through privacy engine too?

File: opacus/tests/accountants_test.py [hunk 3]
```diff
@@ -119,3 +119,59 @@ def test_get_noise_multiplier_gdp(self):
...
+        # check loading logic
+        self.assertNotEqual(new_rdp_accountant.state_dict(), accountant.state_dict())
+        new_rdp_accountant.load_state_dict(accountant.state_dict())
+        self.assertEqual(new_rdp_accountant.state_dict(), accountant.state_dict())
+
+        # ensure correct output after completion
+        for _ in range(steps - 1000, steps):
+            new_rdp_accountant.step(
+                noise_multiplier=noise_multiplier, sample_rate=sample_rate
+            )
```
Comment (ffuuugor) [inline opacus/tests/accountants_test.py:174]: That's a great idea for a test!

**Review by ffuuugor — COMMENTED** [review]
Not surprising at all, but amazing stuff! Thanks for the clean implementation and test coverage!

Slight ask: can we also have a new test in privacy_engine_test for the combined method?

**Comment by karthikprasad** [comment]
Thanks Igor!
> can we also have a new test in privacy_engine_test for the combined method?

Yes, that is still pending, and it the only reason this PR is still in draft mode for this long. I'll add it soon and open it up for review :)

File: opacus/grad_sample/grad_sample_module.py (hunk 1 above)
Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:428]: >  do we want to reroute all unknown (i.e. not explicitly declared in GradSampleModule) attributes to the underlying _module?

Good point! Perhaps not! Also, I realise now that we should also store GS specific attributes such as `loss_reduction` and `batch_first` in the `GS.state_dict()`

>  discrepancy between the output of state_dict and the output of named_parameters()

The only difference right now is the extra `_module` prefix, correct?

Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py:428]: > Good point! Perhaps not!

Or perhaps yes...

> I realise now that we should also store GS specific attributes such as loss_reduction and batch_first in the GS.state_dict()

Should we though? Are those fundamentally part of the model state or model architecture?

> The only difference right now is the extra _module prefix, correct?

Affirmative

**Comment by ffuuugor** [comment]
> Yes, that is still pending, and it the only reason this PR is still in draft mode for this long. I'll add it soon and open it up for review :)

Whoops, sorry I jumped the gun - the PR looked so nice and polished, calling to get reviewed

File: opacus/privacy_engine.py (hunk 2 above)
Comment (karthikprasad) [reply opacus/privacy_engine.py:503]: Re 1: Point well taken.
Re 2: Scheduler already has a `state_dict()` method. I think it definitely makes sense to save that from Privacy Engine too.

File: opacus/grad_sample/grad_sample_module.py (hunk 1 above)
Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:428]: > Should we though? Are those fundamentally part of the model state or model architecture?

They are not, but are essential to instantiate a GS module. I don't see a downside if we do. Do you feel otherwise?

File: docs/faq.md [hunk 4]
```diff
@@ -88,7 +88,7 @@ This statement extends to all downstream uses of this model: its inferences, fin
...
 
-Epsilon and delta are computed *ex post*, following an optimizer run. In fact, for each delta there’s some epsilon, depending on that delta, such that the run satisfies (epsilon, delta)-DP. The call `privacy_engine.get_privacy_spent(delta)` outputs that epsilon in its first return value.
+Epsilon and delta are computed *ex post*, following an optimizer run. In fact, for each delta there’s some epsilon, depending on that delta, such that the run satisfies (epsilon, delta)-DP. The call `privacy_engine.accountant.get_privacy_spent(delta=delta)` outputs that epsilon in its first return value.
```
Comment (ffuuugor) [inline docs/faq.md:91]: nit: recommended way of accessing epsilon now is `privacy_engine.get_epsilon(delta=delta)`

File: opacus/grad_sample/grad_sample_module.py [hunk 5]
```diff
@@ -425,6 +425,35 @@ def rearrange_grad_samples(
...
         return activations, backprops
 
+    def state_dict(self, *args, **kwargs) -> Dict:
+        """
+        Return the state dict of the wrapped module
+        """
+        ret_state_dict = {
+            f"_module.{key}": value
+            for key, value in self._module.state_dict(*args, **kwargs).items()
+        }
```
Comment (ffuuugor) [inline opacus/grad_sample/grad_sample_module.py:435]: Are you sure about this?
In my experience "_module" prefix is always an annoyance and never helps. When I was mentioning the discrepancy in my previous comments, I meant that we should approach it from a different end and remove the prefix everywhere. WDYT?

In particular, getting rid of the prefix makes it much easier to load weights from publicly pre-trained models (currently you have to do it strictly before you wrap a model with GSM)

File: opacus/grad_sample/grad_sample_module.py [hunk 6]
```diff
@@ -425,6 +425,35 @@ def rearrange_grad_samples(
...
+        ret_state_dict["batch_first"] = self.batch_first
+        ret_state_dict["loss_reduction"] = self.loss_reduction
```
Comment (ffuuugor) [inline opacus/grad_sample/grad_sample_module.py:437]: I would vote against having this in the `state_dict`, especially `loss_reduction`.
They are fundamentally part of an architecture, not a state of the model. Additionally, one might want to use old weights stored in a state_dict to train with a different loss, thus changing loss reduction - and I don't think there's anything wrong with that.

`batch_first` if trickier, and I don't have a good use case, but there's always one-to-one mapping between model architecture and batch_first - and we don't store arch in the state_dict

File: opacus/privacy_engine.py [hunk 7]
```diff
@@ -493,3 +495,59 @@ def get_epsilon(self, delta):
...
+        )
+        dict_to_save["privacy_accountant_state_dict"] = self.accountant.state_dict()
+        if optimizer is not None:
+            dict_to_save["optimizer_state_dict"] = optimizer.state_dict()
+        if noise_scheduler is not None:
+            dict_to_save["noise_scheduler_state_dict"] = noise_scheduler.state_dict()
+
+        torch.save(dict_to_save, path, **(torch_save_kwargs or {}))
+
+    def load_checkpoint(
```
Comment (ffuuugor) [inline opacus/privacy_engine.py:531]: This handles default values beautifully, thanks for that!

Question: do we want to provide a warning in any of the following scenarios:
1) optimizer is provided, but checkpoint doesn't have optimzier_state_dict
2) checkpoint have optimizer_state_dict, but optimizer isn't provided

File: opacus/grad_sample/grad_sample_module.py (hunk 5 above)
Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:435]: > in my previous comments, I meant that we should approach it from a different end and remove the prefix everywhere. WDYT?
Ah! That works too. I don't have a strong preference so long as it is consistent. I'll change that.

> In my experience "_module" prefix is always an annoyance and never helps.

Out of curiosity, what trouble ddi the prefix cause in the past>

File: opacus/grad_sample/grad_sample_module.py (hunk 6 above)
Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py:437]: Fair points, both!
The way I thought of this is that these two, though not part of a model's state, are necessary to rebuild the model in the same state as before. If a new opacus user wants to save a checkpoint to be reloaded later, they might not know to save these extra variables. However, this can easily be solved with documentation and/or helpful message.

Alternatively, WDYT about saving this is as part of privacy engine checkpoints?

File: opacus/privacy_engine.py (hunk 7 above)
Comment (karthikprasad) [reply opacus/privacy_engine.py:531]: I don't see why not.

File: opacus/grad_sample/grad_sample_module.py (hunk 5 above)
Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py:435]: Scenario 1: You have a "--disable-dp" flag in your pipeline, and you want to train and compare dp and non-dp model. Imagine also, that your infra is unstable, so you checkpoint the model often and run evaluation in a separate job. You then need to handle two different formats of state_dicts
Scenario 2: In general, when you train DP model and then want to transform it to inference model (non-DP obviously), you'd rather have same state_dicts
Scenario 3: When doing public pretraining, today you have to be careful and load_state_dict before you wrap the model with GradSampleModule. With same state_dicts you don't have to worry about it

File: opacus/grad_sample/grad_sample_module.py (hunk 6 above)
Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py]: >  If a new opacus user wants to save a checkpoint to be reloaded later, they might not know to save these extra variables

Do you consider loading your own checkpoint or loading one from a public hub? 

> Alternatively, WDYT about saving this is as part of privacy engine checkpoints?

I'm not sure I see how it can be loaded in a clear way? Would we override parameters set by GradSampleModule constructor for the module that was passed in?

Comment (karthikprasad) [reply opacus/grad_sample/grad_sample_module.py]: > Do you consider loading your own checkpoint or loading one from a public hub?

Both.

> I'm not sure I see how it can be loaded in a clear way? Would we override parameters set by GradSampleModule constructor for the module that was passed in?

I agree; it is not clear and also not neat. I do think there is value in saving metadata such as these in the checkpoint, but that would mean having to decide what to do at load time, which is not clear. 

To address this, in the current implementation, I have added support for passing your own dict to `save_checkpoint`. This dict can have all the metadata the user thinks is useful to save along with the state_dicts. At load time, we simply load the states and return the entire loaded dict (state_dicts and metadata) to the user for them to use it the way they see fit.

Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py]: > To address this, in the current implementation, I have added support for passing your own dict to save_checkpoint

Oh, that's neat, I love it

File: opacus/grad_sample/grad_sample_module.py (hunk 5 above)
Comment (ffuuugor) [reply opacus/grad_sample/grad_sample_module.py:435]: JFYI, to clarify things, `state_dict` and `load_state_dict` are already available in GradSampleModule today, as it's inherited from nn.Module, and it behaves exactly like this, so there's no need to override the method even. 

To be fair, one argument for leaving this alone for now is:
a) it doesn't directly help the task at hand, checkpointing can work with either options
b) it will break backward compatibility for some use cases, and even though I do believe it's a change for good, we might want to approach this more carefully.

Before moving forward, let's do one of the following:
1) Decide to deal with parameter names later and remove `state_dict` and `load_state_dict` overrides (because default inherited behaviour is identical)
2) Decide to rename parameters in state_dict now and don't add "_module." prefix

I'm fine with either, and for me that's the only outstanding item before accepting

**Review by ffuuugor — APPROVED** [review]
Thanks for the amazing diff and for addressing the comments!
Tests seem to broken in main too, approving
